plugins/hls-explicit-record-fields-plugin/src/Ide/Plugin/ExplicitFields.hs

```diff
@@ -137,23 +147,42 @@ codeActionProvider ideState pId (CodeActionParams _ _ docId range _) = pluginRes
           title = "Expand record wildcard"
 
 collectRecordsRule :: Recorder (WithPriority Log) -> Rules ()
-collectRecordsRule recorder = define (cmapWithPrio LogShake recorder) $ \CollectRecords nfp -> do
-  tmr <- use TypeCheck nfp
-  let exts = getEnabledExtensions <$> tmr
-      recs = concat $ maybeToList (getRecords <$> tmr)
-  logWith recorder Debug (LogCollectedRecords recs)
-  let renderedRecs = traverse renderRecordInfo recs
-      recMap = RangeMap.fromList (realSrcSpanToRange . renderedSrcSpan) <$> renderedRecs
-  logWith recorder Debug (LogRenderedRecords (concat renderedRecs))
-  pure ([], CRR <$> recMap <*> exts)
+collectRecordsRule recorder = define (cmapWithPrio LogShake recorder) $ \CollectRecords nfp ->
+  justOrFail "Unable to TypeCheck" (use TypeCheck nfp) $ \tmr -> do
+    let exts = getEnabledExtensions tmr
+        recs = getRecords tmr
+    logWith recorder Debug (LogCollectedRecords recs)
+    let names = getNames tmr
+        renderedRecs = traverse (renderRecordInfo names) recs
+        recMap = RangeMap.fromList (realSrcSpanToRange . renderedSrcSpan) <$> renderedRecs
+    logWith recorder Debug (LogRenderedRecords (concat renderedRecs))
+    pure ([], CRR <$> recMap <*> Just exts)
+
   where
     getEnabledExtensions :: TcModuleResult -> [GhcExtension]
     getEnabledExtensions = map GhcExtension . getExtensions . tmrParsed
 
+    justOrFail :: MonadFail m => String -> m (Maybe a) -> (a -> m b) -> m b
+    justOrFail = flip . maybeM . fail
+
 getRecords :: TcModuleResult -> [RecordInfo]
 getRecords (tmrRenamed -> (hs_valds -> valBinds,_,_,_)) =
   collectRecords valBinds
 
+-- | Collects all 'Name's of a given source file, to be used
+-- in the variable usage analysis.
```

Comment:
More specifically, it collects all the names and a backwards lookup to the original AST node that the name was harvested from?

Reply:
And it seems like that last bit is mostly because of the srcspan check you do later, which I still don't really understand the need for. Is there a test case that fails without it?

Reply:
It is easier to explain over an example:
```haskell
module Test where

f :: Int -> Int
f foo = foo + 3 * foo + let foo = 5 in foo + 1
```
In this code segment, there are 2 **unique** names (excluding the names not related to `foo`), namely the `foo` that is the function argument and the `foo` which is let-bound. However, there are 5 names in total (again, that occurs as `foo` in the source). GHC lists all these as follows:
```
({ Test.hs:4:3-5 } {Name: foo_atU})
({ Test.hs:4:9-11 } {Name: foo_atU})
({ Test.hs:4:19-21 } {Name: foo_atU})
({ Test.hs:4:29-31 } {Name: foo_awb})
({ Test.hs:4:40-42 } {name: foo_awb})
```
Indeed, this is what we expected. There are 3 references to the `foo` that is the argument to the function, **one of which is the argument itself.** If we are only concerned with whether this argument is used, we shouldn't consider this first `Name`, and check if there are other `Name`s with the same `Unique` (in this example, `foo_atU`). The source span inequality check ensures that we in fact exclude this first `Name` that corresponds to the argument in our considerations.

---

Now, it is obvious that such filtering is necessary. However, considering the fact that we build a map keyed with the `Unique`s of `Name`s, we can simplify our filtering, and probably can just check if the list of `Name`s corresponding to a `Unique` has length greater than 1 (since we know the name at the binding-site is always present). The initial implementation worked with just a big list of `Name`s rather than building a map, and this check involving source spans was carried over from that implementation.

Reply:
Hmmmm. Maybe we're not quite stating the problem right? From your example I would say that the problem is that you are looking for all _uses_ of `foo`, but your naive traversal also turns up _definitions_ of `foo`. I'm kind of surprised that GHC doesn't have some tools for finding uses... but maybe it has a way to tell whether something is a definition? I'm not entirely convinced that your solutions as described will work. e.g. for "list of names greater than 1", consider
```
haha :: Int -> Int -> Int
haha foo 1 = 2
haha foo _ = 3
```
... although maybe that will have different uniques for the different `foo`s? Unclear. That makes me wonder if there's something more robust to be had in GHC. Maybe worth asking the GHC devs.

Reply:
I have asked in GHC chatroom whether GHC provides anything to find the _uses_ of a `Name`, but it doesn't seem to be the case. In the example you provided it isn't an issue because as you pointed out GHC indeed assigns different uniques to those `foo`s. I don't expect there to be an edge-case regarding this approach, since GHC does all the heavy-lifting regarding the name resolution. We merely make use of a single assumption: that given a list of names having same uniques, only a single of them is the defining one. I would be really surprised if there was a case where this does not hold.

Reply:
I think you're probably right, it just feels like a thing that might have nasty edge cases lurking. But if GHC doesn't give us anything (surprisingly?) then this seems good to me!